osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuMovement.cs

```diff
@@ -0,0 +1,631 @@
+                { 1, 1, 1, 1, 1, 1 },
+                { 0.4, 0.4, 0.2, 0.4, 0.2, 0.2 }
+            },
+            {
+                { 0, 0, 0, 0, 0, 0 },
+                { 0, 0, 0, 0, 0, 0 },
+                { 0, 0, 0, 0, 0, 0 },
+                { 0, 0, 1, 0.6, 0.6, 0.6 }
+            },
+            {
+                { 1, 1, 1.5, 2, 3, 3 },
+                { 0, 0, 0, 0, 0, 0 },
+                { 1, 1, 1, 1, 1, 1 },
+                { 0, 0, -0.6, -0.4, -0.3, -0.3 }
+            }
+        };
+
+        private static LinearSpline k0FlowInterp;
+        private static LinearSpline scale0FlowInterp;
+        private static LinearSpline[,] coeffs0FlowInterps;
+        private static LinearSpline k0SnapInterp;
+        private static LinearSpline scale0SnapInterp;
+        private static LinearSpline[,] coeffs0SnapInterps;
+        private static LinearSpline k3FlowInterp;
+        private static LinearSpline scale3FlowInterp;
+        private static LinearSpline[,] coeffs3FlowInterps;
+        private static LinearSpline k3SnapInterp;
+        private static LinearSpline scale3SnapInterp;
+        private static LinearSpline[,] coeffs3SnapInterps;
+
+        private const double t_ratio_threshold = 1.4;
+        private const double correction0_still = 0;
+
+        public double RawMt { get; private set; }
+        public double D { get; private set; }
+        public double Mt { get; private set; }
+        public double Ip12 { get; private set; }
+        public double Cheesablility { get; private set; }
+        public double CheesableRatio { get; private set; }
+        public double Time { get; private set; }
+        public bool EndsOnSlider { get; private set; }
+
+        /// <summary>
+        /// Extracts movement (only for the first object in a beatmap).
+        /// </summary>
+        public static List<OsuMovement> ExtractMovement(OsuHitObject obj)
+        {
+            var movement = GetEmptyMovement(obj.StartTime / 1000.0);
+
+            var movementWithNested = new List<OsuMovement> { movement };
+            // add zero difficulty movements corresponding to slider ticks/slider ends so combo is reflected properly
+            int extraNestedCount = obj.NestedHitObjects.Count - 1;
+
+            for (int i = 0; i < extraNestedCount; i++)
+            {
+                movementWithNested.Add(GetEmptyMovement(movement.Time));
+            }
+
+            return movementWithNested;
+        }
+
+        /// <summary>
+        /// Calculates the movement time, effective distance and other details for the movement from obj1 to obj2.
+        /// </summary>
+        public static List<OsuMovement> ExtractMovement(OsuHitObject obj0, OsuHitObject obj1, OsuHitObject obj2, OsuHitObject obj3,
+                                                        Vector<double> tapStrain, double clockRate,
+                                                        bool hidden = false, double noteDensity = 0, OsuHitObject objMinus2 = null)
```

Comment:
Is this operating on sequences of four consecutive hitobjects, sometimes also taking into account the one before last? What do the indices mean? It's not exactly clear.

Reply:
Added xmldoc, but i think it needs some variable renaming as well but I'm not sure what will make it more readable considering all t01/12/etc vars